### DISCUSSION TITLE
How to configure WinFsp For Case-Insensitive Tab-Completion

### DISCUSSION THREAD

**Discussion by networkproblemsolver** [winfsp/winfsp#428] [General]
Hi, I am using Cryptomator, and recently changed to WinFsp from DokAny.

DokAny would do case-insensitive tab-completion at the command prompt, but WinFsp does not.  If I have a Cryptomator folder named X:\UT and do a "cd u*" and press TAB, I would see a "completion" option of UT... but with WinFsp, tab-completion gives nothing unless I do "cd U*" (with a capital U).

I know I could rename all my files and folders to lower case, but I'm hoping someone here might have an easier way to allow for case-insensitive tab-completion ???

**Comment by billziss-gh** [comment]
This really depends on the file system.

If the file system is case-insensitive then tab completion will work in a case-insensitive manner. If the file system is case-sensitive then tab completion will work in a case-sensitive manner.

@infeo or @overheadhunter should be able to answer the question whether Cryptomator is case-insensitive or not. (The [`FUSE_CAP_CASE_INSENSITIVE`](https://github.com/winfsp/winfsp/blob/v1.10/inc/fuse/fuse_common.h#L50) capability is used to make a WinFsp-FUSE file system case-insensitive if Cryptomator supports this.)

**Reply by networkproblemsolver** [reply]
I am using Windows 10 Pro 21H2 (OS Build 19044.1415), sorry I forgot to mention that.

Cryptomator is case sensitive, unfortunately, but with DokAny v1 I'm able to do the tab-completion without "sensitivity"...

**Reply by billziss-gh** [reply]
Unfortunately if the file system is case-sensitive you will not be able to do tab completion in a case-insensitive manner.

**Reply by networkproblemsolver** [reply]
So how does DOKANY do it ???

**Reply by overheadhunter** [reply]
I guess they do a full dir listing and return the first case-insensitive match. While there is a small performance penalty, it allows tab completion.

That said, without a complete dir listing it is still impossible to access a node directly when the correct case-sensitive name is unknown.

**Reply by networkproblemsolver** [reply]
Ok, so DOKANY v1 is doing a "full dir listing" and then returning a case INSENSITIVE list for both "dir n*" and "n"+TAB.  I wonder if DOKANY v2 will do the same, or if Cryptomator will support v2...

I am guessing that WinFsp will not -- so I'll try one of the other vendor options.

Thx

**Comment by networkproblemsolver** [comment]
Hi Bill, I see you now have the ability to uninstall WinFsp without a reboot, amazing!!

I wanted to follow-up on the "case-insensitive" request... I hope you might give it some more thought.

I checked with the DokAny devs, they will continue to support case-insensitive TABS and DIR in v2 (but Cryptomator is slow to adopt v2).

Yours,
Jim

**Comment by billziss-gh** [comment]
Unfortunately I have no plans to change this.

In a case-sensitive file system the files “ABC” and “abc” are different. So it is semantically incorrect to pretend that they are the same, even if convenient in some cases, like TAB completion.
